### PR TITLE
Specify multiple kernel directories (closes #1250)

### LINKED ISSUE #1250: Look for kernels in more than one location
Currently we can only look into one location for kernels using the `-d` option. It would be useful to be able to point to multiple locations for kernel source.

This should just be a modification to the top level script with an agreed interface.

### PR DESCRIPTION
**Pull request by rupertford**
Adds support for specifying multiple directories when searching for kernel code

### CODE REVIEW

**Comment by TeranIvy** [comment]
I'll take it so I can test it in the LFRic build system.

File: src/psyclone/tests/generator_test.py [hunk 1]
```diff
@@ -98,7 +98,7 @@ def teardown_function():
 
 def test_non_existant_filename():
     ''' checks that alg_gen raises appropriate error when a
-    non-existant filename is supplied '''
+        non-existant filename is supplied '''
```
Comment (TeranIvy) [inline src/psyclone/tests/generator_test.py:101]: non-existent

File: src/psyclone/tests/parse/algorithm_test.py [hunk 2]
```diff
@@ -38,6 +38,7 @@
 tests for code that is not covered there.'''
 
 from __future__ import absolute_import
+import os
```
Comment (TeranIvy) [inline src/psyclone/tests/parse/algorithm_test.py:41]: I am afraid this test module only works when the `pytest` is run from the `src/psyclone/tests/parse` directory where the file is located. It should work regardless of the location when running `pytest`.

File: src/psyclone/tests/parse/kernel_test.py [hunk 3]
```diff
@@ -43,14 +43,23 @@
 import os
 import pytest
 from fparser.api import parse
+from fparser import api as fpapi
```
Comment (TeranIvy) [inline src/psyclone/tests/parse/kernel_test.py:46]: I am afraid this test module only works when the `pytest` is run from the `src/psyclone/tests/parse` directory where the file is located. It should work regardless of the location when running `pytest`.

**Review by TeranIvy — CHANGES_REQUESTED** [review]
The changes do what they say, and that is enabling multiple locations for the kernel source (tested in LFRic). Unfortunately there seem to be problems with running `algorithm_test.py` and `kernel_test.py` from any other directory than where the test modules are located (see inline comments).

I also have some suggestions for documentation and the docstrings, as well as raising exceptions in the modified functions (see inline).

The developer guide documentation builds fine, examples run and all changes are covered. We discussed the current test failures and the test coverage. I would kindly ask the developer to refrain from major test modifications (except for corrections of failures) until the next round of review 😃

File: src/psyclone/tests/parse/algorithm_test.py (hunk 2 above)
Comment (rupertford) [reply src/psyclone/tests/parse/algorithm_test.py:41]: I confirm that I also get the same error - oops.

File: src/psyclone/tests/parse/kernel_test.py (hunk 3 above)
Comment (rupertford) [reply src/psyclone/tests/parse/kernel_test.py:46]: I confirm this error too.

File: doc/user_guide/psyclone_command.rst [hunk 4]
```diff
@@ -236,11 +236,10 @@ specified directory:
     > psyclone -d tests/test_files/dynamo0p3 -api dynamo0.3 use.f90 
     [code output]
 
-.. note::
-    The ``-d`` option is limited to a single directory. Therefore a
-    current limitation in PSyclone is that all kernel files
-    required by an algorithm file must exist within a directory
-    hierarchy where their file names are unique.
+.. note:: The ``-d`` option can be repeated to add as many search
+    directories as is required, with the same constraint as before,
```
[the remark this answers is not in the document]
Comment (rupertford) [reply doc/user_guide/psyclone_command.rst:240]: Updated thanks.

File: src/psyclone/generator.py [hunk 5]
```diff
@@ -152,46 +152,47 @@ def generate(filename, api="", kernel_path="", script_name=None,
...
-                                    default is set in the config.py file.
-    :param str kern_out_path: Directory to which to write transformed \
-                              kernel code.
+    :param str filename: the file containing the algorithm specification.
+    :param kernel_paths: the directories from which to recursively \
+        search for the files containing the kernel source (if \
+        different from the location of the algorithm specification). \
+        Defaults to None.
+    :type kernel_paths: list of str or NoneType
+    :param str script_name: a script file that can apply optimisations \
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/generator.py:161]: Added.

File: src/psyclone/generator.py [hunk 6]
```diff
@@ -152,46 +152,47 @@ def generate(filename, api="", kernel_path="", script_name=None,
...
+        to the PSy layer (can be a path to a file or a filename that \
+        relies on the PYTHONPATH to find the module).
+    :param bool line_length: a logical flag specifying whether we care \
+        about line lengths being longer than 132 characters. If so, \
+        the input (algorithm and kernel) code is checked to make sure \
+        that it conforms. The default is False.
+    :param bool distributed_memory: a logical flag specifying whether \
+        to generate distributed memory code. The default is set in the \
+        config.py file.
+    :param str kern_out_path: directory to which to write transformed \
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/generator.py:171]: And this.

File: src/psyclone/generator.py [hunk 7]
```diff
@@ -152,46 +152,47 @@ def generate(filename, api="", kernel_path="", script_name=None,
...
+        kernel code.
     :param bool kern_naming: the scheme to use when re-naming transformed \
-                             kernels.
+        kernels.
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/generator.py:174]: This too.

File: src/psyclone/generator.py [hunk 8]
```diff
@@ -152,46 +152,47 @@ def generate(filename, api="", kernel_path="", script_name=None,
...
+    :param str script_name: a script file that can apply optimisations \
+        to the PSy layer (can be a path to a file or a filename that \
+        relies on the PYTHONPATH to find the module).
+    :param bool line_length: a logical flag specifying whether we care \
+        about line lengths being longer than 132 characters. If so, \
+        the input (algorithm and kernel) code is checked to make sure \
+        that it conforms. The default is False.
+    :param bool distributed_memory: a logical flag specifying whether \
+        to generate distributed memory code. The default is set in the \
+        config.py file.
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/generator.py:170]: Done.

File: src/psyclone/generator.py [hunk 9]
```diff
@@ -215,12 +216,14 @@ def generate(filename, api="", kernel_path="", script_name=None,
 
     if not os.path.isfile(filename):
         raise IOError("file '{0}' not found".format(filename))
-    if kernel_path and not os.access(kernel_path, os.R_OK):
-        raise IOError("kernel search path '{0}' not found".format(kernel_path))
+    for kernel_path in kernel_paths:
+        if not os.access(kernel_path, os.R_OK):
+            raise IOError(
+                "kernel search path '{0}' not found".format(kernel_path))
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/generator.py:222]: Agreed. Modified this and one above.

File: src/psyclone/parse/algorithm.py [hunk 10]
```diff
@@ -62,32 +62,31 @@
...
-    :param bool line_length: A logical flag specifying whether we care \
-                             about line lengths being longer than 132 \
-                             characters. If so, the input (algorithm \
-                             and kernel) code is checked to make sure \
-                             that it conforms and an error raised if \
-                             not. The default is False.
+    :param str alg_filename: the file containing the algorithm \
+        specification.
+    :param str api: the PSyclone API to use when parsing the code.
+    :param str invoke_name: the expected name of the invocation calls \
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/parse/algorithm.py:75]: Added.

File: src/psyclone/parse/algorithm.py [hunk 11]
```diff
@@ -62,32 +62,31 @@
...
-                            algorithm source).
-    :param bool line_length: A logical flag specifying whether we care \
-                             about line lengths being longer than 132 \
-                             characters. If so, the input (algorithm \
-                             and kernel) code is checked to make sure \
-                             that it conforms and an error raised if \
-                             not. The default is False.
+    :param str alg_filename: the file containing the algorithm \
+        specification.
+    :param str api: the PSyclone API to use when parsing the code.
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/parse/algorithm.py:74]: Documenterised.

File: src/psyclone/generator.py [hunk 12]
```diff
@@ -152,46 +152,47 @@ def generate(filename, api="", kernel_path="", script_name=None,
...
-                             about line lengths being longer than 132 \
-                             characters. If so, the input (algorithm \
-                             and kernel) code is checked to make sure \
-                             that it conforms. The default is False.
-    :param bool distributed_memory: A logical flag specifying whether to \
-                                    generate distributed memory code. The \
-                                    default is set in the config.py file.
-    :param str kern_out_path: Directory to which to write transformed \
-                              kernel code.
+    :param str filename: the file containing the algorithm specification.
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/generator.py:155]: api was not documented at all - eek. 'tis now.

File: src/psyclone/generator.py [hunk 13]
```diff
@@ -152,46 +152,47 @@ def generate(filename, api="", kernel_path="", script_name=None,
...
-                             kernels.
+        kernels.
     :return: 2-tuple containing fparser1 ASTs for the algorithm code and \
-             the psy code.
+        the psy code.
     :rtype: (:py:class:`fparser.one.block_statements.BeginSource`, \
              :py:class:`fparser.one.block_statements.Module`)
 
-    :raises IOError: if the filename or search path do not exist
+    :raises IOError: if the filename or search path do not exist.
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/generator.py:180]: OK :-)

File: src/psyclone/tests/generator_test.py (hunk 1 above)
Comment (rupertford) [reply src/psyclone/tests/generator_test.py:101]: I think I've always spelled that incorrectly. I shall try not to from now on and have replaced all occurrences in PSyclone (of which there were quite a few).

File: src/psyclone/generator.py [hunk 14]
```diff
@@ -138,7 +138,7 @@ def handle_script(script_name, psy):
         os.sys.path.pop()
 
 
-def generate(filename, api="", kernel_path="", script_name=None,
+def generate(filename, api="", kernel_paths=None, script_name=None,
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/generator.py:150]: Sorted these and the others in this file.

**Comment by rupertford** [comment]
Ready for next look from @TeranIvy

File: src/psyclone/generator.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (TeranIvy) [reply src/psyclone/generator.py:150]: Thanks! I'mhappy to report the perfect `pylint` score for this module now 😄

File: src/psyclone/tests/generator_test.py (hunk 1 above)
Comment (TeranIvy) [reply src/psyclone/tests/generator_test.py]: Thank you for changing them!

File: src/psyclone/tests/parse/kernel_test.py (hunk 3 above)
Comment (TeranIvy) [reply src/psyclone/tests/parse/kernel_test.py:46]: Thanks for fixing, all works now from the top working copy directory and from the `src/psyclone/tests` directory.

File: src/psyclone/tests/parse/algorithm_test.py (hunk 2 above)
Comment (TeranIvy) [reply src/psyclone/tests/parse/algorithm_test.py:41]: All working now, thanks for fixing 😄

File: src/psyclone/tests/parse/kernel_test.py [hunk 15]
```diff
@@ -168,11 +179,48 @@ def test_getkernelfilepath_multifile(tmpdir):
...
+
+    '''
+    kern_module_name = "testkern_mod"
+    alg_file_name = os.path.join(LFRIC_BASE_PATH, "1_single_invoke.f90")
+    result = get_kernel_filepath(kern_module_name, [], alg_file_name)
+    assert "testkern_mod.F90" in result
+
+
+def test_getkernelfilepath_nomatch():
+    '''Test that the expected exception is raised if the kernel file is
```
Comment (TeranIvy) [inline src/psyclone/tests/parse/kernel_test.py:199]: Blank space before the text, please.

File: src/psyclone/tests/parse/kernel_test.py [hunk 16]
```diff
@@ -168,11 +179,48 @@ def test_getkernelfilepath_multifile(tmpdir):
...
+    kern_module_name = "testkern_mod"
+    alg_file_name = os.path.join(LFRIC_BASE_PATH, "1_single_invoke.f90")
+    with pytest.raises(ParseError) as info:
+        get_kernel_filepath(
+            kern_module_name, [GOCEAN_BASE_PATH], alg_file_name)
+    assert ("Kernel file 'testkern_mod.[fF]90' not found in"
+            in str(info.value))
+
+
+def test_getkernelfilepath_multidir():
```
Comment (TeranIvy) [inline src/psyclone/tests/parse/kernel_test.py:212]: Blank space before the text, please.

File: src/psyclone/tests/parse/kernel_test.py [hunk 17]
```diff
@@ -168,11 +179,48 @@ def test_getkernelfilepath_multifile(tmpdir):
...
+    '''
+    kern_module_name = "testkern_mod"
+    alg_file_name = os.path.join(LFRIC_BASE_PATH, "1_single_invoke.f90")
+    result = get_kernel_filepath(
+        kern_module_name, [GOCEAN_BASE_PATH, LFRIC_BASE_PATH], alg_file_name)
+    assert "testkern_mod.F90" in result
+
+
 def test_getkernelfilepath_caseinsensitive1(tmpdir):
     '''Test that a case insensitive match is performed when searching for
```
Comment (TeranIvy) [inline src/psyclone/tests/parse/kernel_test.py:225]: Blank space before the text, please.

File: src/psyclone/tests/parse/kernel_test.py [hunk 18]
```diff
@@ -202,10 +250,52 @@ def test_getkernelfilepath_caseinsensitive2(tmpdir):
...
-    result = get_kernel_filepath("TEST_MOD", None, filename)
+    result = get_kernel_filepath("TEST_MOD", [], filename)
     assert "tmp" in result
     assert "test_mod.f90" in result
 
+# function get_kernel_ast
+
+
+def test_getkernelast_nodir():
+    '''Test that the directory of the algorithm file is searched if no
```
Comment (TeranIvy) [inline src/psyclone/tests/parse/kernel_test.py:261]: Blank space before the text, please.

File: src/psyclone/tests/parse/kernel_test.py [hunk 19]
```diff
@@ -202,10 +250,52 @@ def test_getkernelfilepath_caseinsensitive2(tmpdir):
...
+
+    '''
+    kern_module_name = "testkern_mod"
+    alg_file_name = os.path.join(LFRIC_BASE_PATH, "1_single_invoke.f90")
+    result = get_kernel_ast(kern_module_name, alg_file_name, [], False)
+    assert isinstance(result, BeginSource)
+
+
+def test_getkernelast_nomatch():
+    '''Test that the expected exception is raised if the kernel file is
```
Comment (TeranIvy) [inline src/psyclone/tests/parse/kernel_test.py:272]: Blank space before the text, please.

File: src/psyclone/tests/parse/kernel_test.py [hunk 20]
```diff
@@ -202,10 +250,52 @@ def test_getkernelfilepath_caseinsensitive2(tmpdir):
...
+    alg_file_name = os.path.join(LFRIC_BASE_PATH, "1_single_invoke.f90")
+    with pytest.raises(ParseError) as info:
+        get_kernel_ast(
+            kern_module_name, alg_file_name, [GOCEAN_BASE_PATH], False)
+    assert ("Kernel file 'testkern_mod.[fF]90' not found in"
+            in str(info.value))
+
+
+def test_getkernelast_multidir():
+    '''Test that get_kernel_ast works when multiple directories are
```
Comment (TeranIvy) [inline src/psyclone/tests/parse/kernel_test.py:286]: Blank space before the text, please.

File: src/psyclone/tests/parse/kernel_test.py [hunk 21]
```diff
@@ -635,7 +721,7 @@ def test_get_integer_array():
 def test_get_int_array_name_err(monkeypatch):
     ''' Tests that we raise the correct error if there is something wrong
     with the Name in the assignment statement obtained from fparser2. '''
-    from fparser.two import Fortran2003
+
```
Comment (TeranIvy) [inline src/psyclone/tests/parse/kernel_test.py:724]: I would suggest a blank line before the closing quotations instead as in other tests.

File: src/psyclone/tests/parse/kernel_test.py [hunk 22]
```diff
@@ -664,7 +750,7 @@ def my_init(self, _):
 def test_get_int_array_constructor_err(monkeypatch):
     ''' Check that we raise the appropriate error if we fail to parse the
     array constructor expression. '''
-    from fparser.two import Fortran2003
+
```
Comment (TeranIvy) [inline src/psyclone/tests/parse/kernel_test.py:753]: I would suggest a blank line before the closing quotations instead as in other tests.

File: src/psyclone/tests/parse/kernel_test.py [hunk 23]
```diff
@@ -690,7 +776,7 @@ def my_init(self, _):
 def test_get_int_array_section_subscript_err(monkeypatch):
     ''' Check that we raise the appropriate error if the parse tree for the
     LHS of the array declaration is broken. '''
-    from fparser.two import Fortran2003
+
```
Comment (TeranIvy) [inline src/psyclone/tests/parse/kernel_test.py:779]: I would suggest a blank line before the closing quotations instead as in other tests.

**Review by TeranIvy — CHANGES_REQUESTED** [review]
Almost there - thanks for improving the code!

Regarding the testing, I think it is better to have more than fewer tests so I am inclined to leave all of them in. I can confirm that the modified tests now work from multiple locations. I have some formatting requests - see inline.

I also have a request for updating the help message for `-d` option - the current one still implies one kernel path (see inline).

The changes are `pycodestyle` and `pylint` clean and fully covered. Documentation builds and the examples run.

The branch also has conflicts with the master that I would kindly ask the developer to resolve.

File: src/psyclone/generator.py [hunk 24]
```diff
@@ -3,7 +3,7 @@
 # -----------------------------------------------------------------------------
 # BSD 3-Clause License
 #
-# Copyright (c) 2017-2020, Science and Technology Facilities Council.
+# Copyright (c) 2017-2021, Science and Technology Facilities Council.
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/generator.py:6]: Updated.

File: doc/user_guide/psyclone_command.rst [hunk 25]
```diff
@@ -236,11 +236,10 @@ specified directory:
     > psyclone -d tests/test_files/dynamo0p3 -api dynamo0.3 use.f90 
     [code output]
 
-.. note::
-    The ``-d`` option is limited to a single directory. Therefore a
-    current limitation in PSyclone is that all kernel files
-    required by an algorithm file must exist within a directory
-    hierarchy where their file names are unique.
+.. note:: The ``-d`` option can be repeated to add as many search
```
[the remark this answers is not in the document]
Comment (rupertford) [reply doc/user_guide/psyclone_command.rst:239]: Updated.

File: src/psyclone/tests/generator_test.py [hunk 26]
```diff
@@ -111,63 +111,70 @@ def test_invalid_api():
...
-                    kernel_path=os.path.join(BASE_PATH,
-                                             "dynamo0p1", "kernels"))
+                    kernel_paths=[
+                        os.path.join(
+                            BASE_PATH, "dynamo0p1", "kernels3", "dead_end"),
+                        os.path.join(BASE_PATH, "dynamo0p1", "kernels")])
 
 
-def test_same_kernel_path():
+def test_same_kernel_paths():
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/tests/generator_test.py:151]: Done

[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/tests/generator_test.py:151]: Done

File: src/psyclone/tests/generator_test.py [hunk 27]
```diff
@@ -111,63 +111,70 @@ def test_invalid_api():
...
     ''' checks that the generator succeeds when the search directory
         is the same as the algorithm code directory and a path is
         specified '''
     path = os.path.join(BASE_PATH, "dynamo0p1", "algorithm")
     _, _ = generate(os.path.join(path, "1_single_function.f90"),
-                    api="dynamo0.1", kernel_path=path)
+                    api="dynamo0.1", kernel_paths=[path])
 
 
 def test_similar_kernel_name():
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/tests/generator_test.py:160]: Done

File: src/psyclone/tests/generator_test.py [hunk 28]
```diff
@@ -111,63 +111,70 @@ def test_invalid_api():
...
                                  "algorithm", "1_single_function.f90"),
                     api="dynamo0.1",
-                    kernel_path=os.path.join(BASE_PATH,
-                                             "dynamo0p1", "kernels2"))
+                    kernel_paths=[
+                        os.path.join(BASE_PATH, "dynamo0p1", "kernels2")])
 
 
-def test_recurse_correct_kernel_path():
+def test_recurse_correct_kernel_paths():
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/tests/generator_test.py:169]: Done

File: src/psyclone/tests/generator_test.py [hunk 29]
```diff
@@ -528,6 +535,21 @@ def test_main_api():
     assert Config.get().api == "dynamo0.1"
 
 
+def test_main_directory_arg(capsys):
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/tests/generator_test.py:538]: Done

File: src/psyclone/tests/generator_test.py [hunk 30]
```diff
@@ -96,11 +96,11 @@ def teardown_function():
 # a set of unit tests for the generate function
 
 
-def test_non_existant_filename():
+def test_non_existent_filename():
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/tests/generator_test.py:99]: Done

File: src/psyclone/tests/parse/algorithm_test.py [hunk 31]
```diff
@@ -47,18 +48,68 @@
...
+                                os.path.pardir, "test_files", "gocean1p0")
+
 # This ParserFactory call needs to happen at the top-level in order for the
 # fparser.two.Fortran2003 import to work as expected.
 ParserFactory().create(std="f2008")
 
+# function parse() tests
+
+
+def test_parse_kernel_paths():
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/tests/parse/algorithm_test.py:69]: Done

File: src/psyclone/tests/generator_test.py (hunk 30 above)
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/tests/generator_test.py:99]: I also changed all other descriptions in this file.

File: src/psyclone/tests/parse/algorithm_test.py [hunk 32]
```diff
@@ -47,18 +48,68 @@
...
+            in str(info.value))
+    # multiple kernel paths
+    parse(alg_name, api="dynamo0.3", kernel_paths=[
+        LFRIC_BASE_PATH, GOCEAN_BASE_PATH])
+
 # class parser() tests
 
 
+def test_parser_init_kernel_paths():
+    '''Check that the Parser class stores the kernel_paths optional
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/tests/parse/algorithm_test.py:94]: Done.

File: src/psyclone/tests/parse/algorithm_test.py [hunk 33]
```diff
@@ -47,18 +48,68 @@
...
+
+    '''
+    alg_name = os.path.join(LFRIC_BASE_PATH, "1_single_invoke.f90")
+    # No argument
+    parse(alg_name, api="dynamo0.3")
+    # None argument
+    parse(alg_name, api="dynamo0.3", kernel_paths=None)
+    # Empty list
+    parse(alg_name, api="dynamo0.3", kernel_paths=[])
+    # invalid path
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/tests/parse/algorithm_test.py:81]: Done.

File: src/psyclone/tests/parse/algorithm_test.py [hunk 34]
```diff
@@ -47,18 +48,68 @@
...
+    with pytest.raises(ParseError) as info:
+        parse(alg_name, api="dynamo0.3", kernel_paths=["invalid"])
+    assert ("Supplied kernel search path does not exist or cannot be read"
+            in str(info.value))
+    # multiple kernel paths
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/tests/parse/algorithm_test.py:86]: Done

File: src/psyclone/tests/parse/algorithm_test.py [hunk 35]
```diff
@@ -47,18 +48,68 @@
...
+    # No argument
+    parser = Parser()
+    assert parser._kernel_paths == []
+    # None argument
+    parser = Parser(kernel_paths=None)
+    assert parser._kernel_paths == []
+    # Empty list
+    parser = Parser(kernel_paths=[])
+    assert parser._kernel_paths == []
+    # multiple kernel paths
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/tests/parse/algorithm_test.py:107]: Done

File: src/psyclone/tests/parse/algorithm_test.py [hunk 36]
```diff
@@ -108,6 +159,37 @@ def test_parser_createinvokecall_error():
         "Expecting argument to be of the form 'name=xxx' or a Kernel call "
         "but found '0.0' in file 'None'.") in str(excinfo.value)
 
+# create_coded_kernel_call tests
+
+
+def test_parser_codedkernelcall_kernel_paths():
+    '''Check that the Parser class passes the kernel_paths information
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/tests/parse/algorithm_test.py:405]: Done.

File: src/psyclone/tests/parse/algorithm_test.py [hunk 37]
```diff
@@ -108,6 +159,37 @@ def test_parser_createinvokecall_error():
...
+    alg_filename = os.path.join(LFRIC_BASE_PATH, "1_single_invoke.f90")
+    parse_tree = parse_fp2(alg_filename)
+    invoke_call = parse_tree.children[0].children[2].children[0]
+    invoke_argument = invoke_call.children[1].children[0]
+    kernel_name, args = get_kernel(invoke_argument, alg_filename)
+    parser = Parser()
+    parser._alg_filename = alg_filename
+    use_statement = parse_tree.children[0].children[1].children[2]
+    parser.update_arg_to_module_map(use_statement)
+    # no paths
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/tests/parse/algorithm_test.py:419]: Done

File: src/psyclone/tests/parse/algorithm_test.py [hunk 38]
```diff
@@ -108,6 +159,37 @@ def test_parser_createinvokecall_error():
...
+    parser.create_coded_kernel_call(kernel_name, args)
+    # invalid kernel path
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/tests/parse/algorithm_test.py:421]: Done

File: src/psyclone/tests/parse/algorithm_test.py [hunk 39]
```diff
@@ -108,6 +159,37 @@ def test_parser_createinvokecall_error():
...
+    parser._kernel_paths = ["invalid"]
+    with pytest.raises(ParseError) as info:
+        parser.create_coded_kernel_call(kernel_name, args)
+    assert ("Supplied kernel search path does not exist or cannot be read"
+            in str(info.value))
+    # multiple kernel paths
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/tests/parse/algorithm_test.py:427]: Done

File: src/psyclone/tests/parse/kernel_test.py [hunk 40]
```diff
@@ -168,11 +179,48 @@ def test_getkernelfilepath_multifile(tmpdir):
...
 
     with pytest.raises(ParseError) as excinfo:
-        _ = get_kernel_filepath("test_mod", str(tmpdir), None)
+        _ = get_kernel_filepath("test_mod", [str(tmpdir)], None)
     assert ("More than one match for kernel file 'test_mod.[fF]90' "
             "found!") in str(excinfo.value)
 
 
+def test_getkernelfilepath_nodir_supplied():
+    '''Test that the directory of the algorithm file is searched if no
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/tests/parse/kernel_test.py:188]: Done

File: src/psyclone/tests/parse/kernel_test.py (hunk 15 above)
Comment (rupertford) [reply src/psyclone/tests/parse/kernel_test.py:199]: Done

File: src/psyclone/tests/parse/kernel_test.py (hunk 16 above)
Comment (rupertford) [reply src/psyclone/tests/parse/kernel_test.py:212]: Done

File: src/psyclone/tests/parse/kernel_test.py (hunk 17 above)
Comment (rupertford) [reply src/psyclone/tests/parse/kernel_test.py:225]: Done

File: src/psyclone/tests/parse/kernel_test.py (hunk 18 above)
Comment (rupertford) [reply src/psyclone/tests/parse/kernel_test.py:261]: Done

File: src/psyclone/tests/parse/kernel_test.py (hunk 19 above)
Comment (rupertford) [reply src/psyclone/tests/parse/kernel_test.py:272]: Done

File: src/psyclone/tests/parse/kernel_test.py (hunk 20 above)
Comment (rupertford) [reply src/psyclone/tests/parse/kernel_test.py:286]: Done

File: src/psyclone/tests/parse/kernel_test.py (hunk 21 above)
Comment (rupertford) [reply src/psyclone/tests/parse/kernel_test.py:724]: Done

File: src/psyclone/tests/parse/kernel_test.py (hunk 22 above)
Comment (rupertford) [reply src/psyclone/tests/parse/kernel_test.py:753]: Done

File: src/psyclone/tests/parse/kernel_test.py (hunk 23 above)
Comment (rupertford) [reply src/psyclone/tests/parse/kernel_test.py:779]: Done

**Comment by rupertford** [comment]
Hopefully addressed reviewer's comments and have brought branch up to master. Ready for next review from @TeranIvy

File: src/psyclone/tests/generator_test.py (hunk 30 above)
[the remark this answers is not in the document]
Comment (TeranIvy) [reply src/psyclone/tests/generator_test.py:102]: Thanks!

**Review by TeranIvy — APPROVED** [review]
All well now. Re-checked and tested everything and re-built documentation.

There was a typo so I took the liberty of fixing it. I also tidied up the indentation in the changelog as discussed with other developers.

Updated changelog and user guide. Awaiting for final tests before merging.

**Pull request merged**